### PR TITLE
Added support for colon (':') in environment variable name

### LINKED ISSUE #53201: Environment variables containing colon not allowed
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Problem:**
Kubernetes does not support `:` in environment variables. I can see that the subject has recently been discussed in [PR #48986](https://github.com/kubernetes/kubernetes/pull/48986). .NET Core uses `:` to replace config variables as mentioned in [this comment](https://github.com/kubernetes/kubernetes/pull/48986#issuecomment-318628326). The limitation in kubernetes requires me to perform some ugly hacks to override the config values.

**Proposed Solution:**
I'm in favor of the sentiment that kubernetes should not censor environment variables at all, but specifically I would like `:` to be allowed.

### PR DESCRIPTION
**Pull request by bsteciuk**
**What this PR does / why we need it**:
This PR adds support for including a colon `:` in environment variable names.  A number of .Net applications/libraries use `:` in their environment variable names.  
**Which issue(s) this PR fixes**:
Fixes #53201 

**Special notes for your reviewer**:

**Release note**:

```release-note
Add support for `:` character in environment variable names.
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go [hunk 1]
```diff
@@ -302,7 +302,7 @@ func IsHTTPHeaderName(value string) []string {
 	return nil
 }
 
-const envVarNameFmt = "[-._a-zA-Z][-._a-zA-Z0-9]*"
+const envVarNameFmt = "[-:._a-zA-Z][-:._a-zA-Z0-9]*"
 const envVarNameFmtErrMsg string = "a valid environment variable name must consist of alphabetic characters, digits, '_', '-', or '.', and must not start with a digit"
```
Comment (michmike) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go]: can we also change the message to include the fact that we now support : ?

Comment (bsteciuk) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go]: Good catch! Updated.

**Review by michmike — APPROVED** [review]

**Comment by PatrickLang** [comment]
Should we consider this for 1.9? Does anyone need that before 1.10 is released?

**Comment by jdumars** [comment]
@bsteciuk you're going to need to contact folks from the OWNERS files listed above and get approvals.

**Review by kow3ns — CHANGES_REQUESTED** [review]
I don't think we should do this. In Linux ":" is a path separator and completely invalid to use in an environment variable name. In .NET Core you use "__" in place of ":" in order get the desired configuration functionality to work [as documented](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/index?tabs=basicconfiguration#simple-configuration)

**Comment by timoreimann** [comment]
@kow3ns FWIW, the argument brought forward in #48986 to support relaxing the rules for dots and dashes was that the mentioned restrictions seemingly only apply to shell contexts; general application usages are supposed to be handled more liberally according to the [standard](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html) (**emphasis** by me):

>Environment variable names used by the utilities in the Shell and Utilities volume of IEEE Std 1003.1-2001 consist solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in Portable Character Set and do not begin with a digit. **Other characters may be permitted by an implementation; applications shall tolerate the presence of such names.** Uppercase and lowercase letters shall retain their unique identities and shall not be folded together. The name space of environment variable names containing lowercase letters is reserved for applications. Applications can define any environment variables with names from this name space without modifying the behavior of the standard utilities.

The referenced [Portable Character Set](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap06.html#tagtcjh_3) lists the colon.

See also the links I noted down in [that PR's description](https://github.com/kubernetes/kubernetes/pull/48986#issue-130732072).

**Comment by bsteciuk** [comment]
Closing due to response and ticket close of #53201, added docs PR to explain the issue and it's workaround kubernetes/website/pull/7657 .

**Pull request closed without merging**